### PR TITLE
Refactor options and examples

### PR DESCRIPTION
**Pull request by zpqrtbnk**
Depends on: #368 

Following discussions with Jaromir, this refactors how options are prepared and how examples are packaged, so that each example class can be taken to a new project and run independently. Also, documentation for options and examples is updated.

### CODE REVIEW

File: src/Hazelcast.Net.Examples/Program.cs [hunk 1]
```diff
@@ -113,7 +113,7 @@ private static string ClientExceptionCode(Exception e)
         private static void InitializeUnobservedExceptions()
         {
             // make sure the queue is empty
-            while (_unobservedExceptions.TryDequeue(out _))
+            while (UnobservedExceptions.TryDequeue(out _))
```
Comment (mdumandag) [inline src/Hazelcast.Net.Examples/Program.cs:116]: Can we use https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentqueue-1.clear?view=net-5.0#System_Collections_Concurrent_ConcurrentQueue_1_Clear here?

File: src/Hazelcast.Net.Examples/Program.cs [hunk 2]
```diff
@@ -136,12 +136,12 @@ private static void CollectUnobservedExceptions()
             GC.WaitForPendingFinalizers();
 
             // check for unobserved exceptions and report
-            var failed = false;
-            while (_unobservedExceptions.TryDequeue(out var args))
+            //var failed = false;
+            while (UnobservedExceptions.TryDequeue(out _))
```
Comment (mdumandag) [inline src/Hazelcast.Net.Examples/Program.cs:140]: There is a comment above saying check exceptions and report. Should we do it here? As far as I see, we are just clearing the queue here, do not report anything back to the user

File: src/Hazelcast.Net/HazelcastOptionsBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+            return this;
+        }
+
+        /// <summary>
+        /// Sets the path (without filename) to the options files to read when building the options.
+        /// </summary>
+        /// <param name="filePath">The file path.</param>
+        /// <returns>This options builder.</returns>
+        /// <remarks>
+        /// <para>By default, when not provided, the options file is searched in searched in the
```
Comment (mdumandag) [inline src/Hazelcast.Net/HazelcastOptionsBuilder.cs:67]: `searched in searched in the` -> `searched in the`

File: doc/dev/doc/configuration.md [hunk 4]
```diff
@@ -113,17 +150,17 @@ services.Configure<HazelcastOptions>(options =>
 });
 ```
 
+Note: the required extension methods are not part of the Hazelcast.Net NuGet packages, but are provided as part of the Hazelcast.Net.DependencyInjection project which is only provided in [source form](https://github.com/hazelcast/hazelcast-csharp-client/tree/master/src/Hazelcast.Net.DependencyInjection).
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply doc/dev/doc/configuration.md:153]: They have to build it for now, but I assume we'll distribute it as NuGet package soon-ish. Just would like it to be more tested by users & gather feedback beforehand.

File: src/Hazelcast.Net.Examples/Client/ClientStatisticsExample.cs [hunk 5]
```diff
@@ -1,61 +0,0 @@
...
-using System;
-using System.Threading;
-using System.Threading.Tasks;
-using Hazelcast.Core;
-using Hazelcast.NearCaching;
-
-namespace Hazelcast.Examples.Client
-{
-    // ReSharper disable once UnusedMember.Global
-    public class ClientStatisticsExample
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply src/Hazelcast.Net.Examples/Client/ClientStatisticsExample.cs:24]: Statistics were never implemented on the C# client so the example actually does nothing and that could be misleading, we'll re-introduce the example when we actually do have statistics.

File: src/Hazelcast.Net.Examples/DotNet/UnobservedExceptionExample.cs [hunk 6]
```diff
@@ -19,9 +19,9 @@
 namespace Hazelcast.Examples.DotNet
 {
     // ReSharper disable once UnusedMember.Global
-    public class UnobservedExceptionExample : ExampleBase
+    public class UnobservedExceptionExample
     {
-        public void Run()
+        public static void Main()
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply src/Hazelcast.Net.Examples/DotNet/UnobservedExceptionExample.cs:24]: This example exists only to show / verify that the mechanism to detect "unobserved exceptions" ie exceptions that happen in background tasks and are never handled (and are a bad thing) actually works. If any example leaks such an exception, the example runner will flag it.

But we don't plan to have more dotnet examples.

File: doc/dev/doc/configuration.md [hunk 7]
```diff
@@ -133,5 +170,6 @@ Host.CreateDefaultBuilder(args)
     });
 ```
 
-Configuration keys will be gathered from the same sources and in the same order as before, and options as well as the 
-@Hazelcast.HazelcastClientFactory will be registered in the service container, and available via dependency injection.
+Just as with the previous container environment, configuration keys will be gathered from the same sources and in the same order as before, and options will be registered in the service container, and available via dependency injection
+
+Note: the required extension methods are not part of the Hazelcast.Net NuGet packages, but are provided as part of the Hazelcast.Net.DependencyInjection project which is only provided in [source form](https://github.com/hazelcast/hazelcast-csharp-client/tree/master/src/Hazelcast.Net.DependencyInjection).
```
[the remark this answers is not in the document]
Comment (zpqrtbnk) [reply doc/dev/doc/configuration.md:175]: ok

File: src/Hazelcast.Net/HazelcastOptionsBuilder.cs (hunk 3 above)
Comment (zpqrtbnk) [reply src/Hazelcast.Net/HazelcastOptionsBuilder.cs:67]: ok

File: src/Hazelcast.Net.Examples/Program.cs (hunk 2 above)
Comment (zpqrtbnk) [reply src/Hazelcast.Net.Examples/Program.cs:140]: we used to, and that is leftover code. removing.

File: src/Hazelcast.Net.Examples/Client/ClientStatisticsExample.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mdumandag) [reply src/Hazelcast.Net.Examples/Client/ClientStatisticsExample.cs:24]: Oh okay, I didn't know that

File: src/Hazelcast.Net.Examples/Program.cs (hunk 1 above)
Comment (zpqrtbnk) [reply src/Hazelcast.Net.Examples/Program.cs:116]: yes, though that is only available with .NET Core and .NET, not .NET Framework.

**Comment by zpqrtbnk** [comment]
verify

**Review by mdumandag — APPROVED** [review]

**Pull request merged**